src/hooks.ts

```diff
@@ -6,6 +6,14 @@ import { Counter, Gauge, Histogram, LabelValues } from 'prom-client';
 import { convertMsToS, filterLabels } from './helpers';
 import { ContextTypes, FieldTypes, MetricsNames, Metrics, MetricTypes } from './metrics';
 
+const CLIENT_ERRORS = [
+  'UNAUTHENTICATED',
+  'BAD_REQUEST',
+  'BAD_USER_INPUT',
+  'GRAPHQL_VALIDATION_FAILED',
+  'GRAPHQL_PARSE_FAILED'
+];
```

Comment:
I think we should only ignore `BAD_USER_INPUT` and `UNAUTHENTICATED` the rest should be count as an error which need to take a look, and fix it. 
any thought on this @snehithv

Reply:
I don't think we should ignore them. The clients should not be making `UNAUTHENTICATED` or `BAD_USER_INPUT`, something wrong with the front end code.

Or someone is trying to attack using token scraping.

Reply:
GRAPHQL_PARSE_FAILED, GRAPHQL_VALIDATION_FAILED : cause by developer's mistake in Frontend side

Reply:
How about we separate to 2 kind of error metrics
One for `client error` and another for `server error`. It more makes sense

Reply:
All coding mistakes need to be monitored, and with the testing which had been done, these errors can and mostly be caused by backward compatibility

Reply:
what does client or server error  mean though ?  Here is the list of `MetricsNames` which is relevant to the code in this PR.

```
export enum MetricsNames {
  SERVER_STARTING = 'apollo_server_starting',
  SERVER_CLOSING = 'apollo_server_closing',
  QUERY_STARTED = 'apollo_query_started',
  QUERY_FAILED = 'apollo_query_failed',
  QUERY_PARSE_STARTED = 'apollo_query_parse_started',
  QUERY_PARSE_FAILED = 'apollo_query_parse_failed',
  QUERY_VALIDATION_STARTED = 'apollo_query_validation_started',
  QUERY_VALIDATION_FAILED = 'apollo_query_validation_failed',
  QUERY_RESOLVED = 'apollo_query_resolved',
  QUERY_EXECUTION_STARTED = 'apollo_query_execution_started',
  QUERY_EXECUTION_FAILED = 'apollo_query_execution_failed',
  QUERY_DURATION = 'apollo_query_duration',
  QUERY_FIELD_RESOLUTION_DURATION = 'apollo_query_field_resolution_duration'
}
```

Reply:
Ah. I mean I will create new metrics to adapt

Reply:
@hoang-tillpos  we do ignores for these cases because of : client may enter the same Product Name, Duplicated Page... which are more about customer input.

Reply:
@quangthanh0909 , a @hoang-tillpos 

Maybe we need to escalate our conversation to the platform group to address the following: 

1. Improve Client-Side Validation
2. Improve Authentication Error Handling
3. Improve API Error Responses
4. Educate Users
5. Add More Context to Monitoring

I think for now we should consider adding more labels to your Prometheus metrics to distinguish between different types of errors. You may want to revise how you calculate your error rate. Since BAD_USER_INPUT and UNAUTHENTICATED errors are often out of your control, you might want to exclude them from your error rate calculation or track them separately from server-side errors.

I add a snippet code here as a reference, @andy-oolio , @quangthanh0909 , let me know what you think: 

`const { Counter } = require('prom-client');

const badUserInputCounter = new Counter({
  name: 'graphql_bad_user_input_errors_total',
  help: 'Total number of GraphQL BAD_USER_INPUT errors',
  labelNames: ['operation_name', 'operation_type'],
});

const unauthenticatedCounter = new Counter({
  name: 'graphql_unauthenticated_errors_total',
  help: 'Total number of GraphQL UNAUTHENTICATED errors',
  labelNames: ['operation_name', 'operation_type'],
});

module.exports = {
  badUserInputCounter,
  unauthenticatedCounter,
  ApolloPrometheusPlugin: {
    requestDidStart(requestContext) {
      return {
        didEncounterErrors(requestContext) {
          const { operationName, operation } = requestContext;

          if (operation) {
            const operationType = operation.operation;
            const labels = { operation_name: operationName, operation_type: operationType };

            requestContext.errors.forEach((error) => {
              if (error.extensions && error.extensions.code === 'BAD_USER_INPUT') {
                badUserInputCounter.inc(labels);
              } else if (error.extensions && error.extensions.code === 'UNAUTHENTICATED') {
                unauthenticatedCounter.inc(labels);
              }
            });
          }
        },
      };
    },
  },
};
`

Reply:
I also push another commit to show my idea.

Reply:
why is the default documentation completely ignored?  did you read the documentation on these?

If you don't want to see the validation metrics, build a dashboard for: `apollo_query_execution_failed` with this formula:
```
1 - (sum(increase(apollo_query_execution_failed{namespace=~"$namespace", service=~"${service}", operation=~"${operation}"}[$__range]) OR vector(0)) by (service) / sum(increase(apollo_query_resolved{namespace=~"${namespace}", service=~"${service}", operation=~"${operation}"}[$__range])) by (service))
```



```
export enum MetricsNames {
  SERVER_STARTING = 'apollo_server_starting',
  SERVER_CLOSING = 'apollo_server_closing',
  QUERY_STARTED = 'apollo_query_started',
  QUERY_FAILED = 'apollo_query_failed',
  QUERY_PARSE_STARTED = 'apollo_query_parse_started',
  QUERY_PARSE_FAILED = 'apollo_query_parse_failed',
  QUERY_VALIDATION_STARTED = 'apollo_query_validation_started',
  QUERY_VALIDATION_FAILED = 'apollo_query_validation_failed',
  QUERY_RESOLVED = 'apollo_query_resolved',
  QUERY_EXECUTION_STARTED = 'apollo_query_execution_started',
  QUERY_EXECUTION_FAILED = 'apollo_query_execution_failed',
  QUERY_DURATION = 'apollo_query_duration',
  QUERY_FIELD_RESOLUTION_DURATION = 'apollo_query_field_resolution_duration'
}
```

Reply:
I don't know why you use custom Apollo plugin hooks to generate `MetricsNames` instead of leaving it to the default plugin. But there are multiple hooks which you can generate the errors, did you go through those hooks ? I can see rightaway 3 hooks:
[didEncounterErrors](https://www.apollographql.com/docs/apollo-server/integrations/plugins-event-reference/#didencountererrors)
[invalidRequestWasReceived](https://www.apollographql.com/docs/apollo-server/integrations/plugins-event-reference/#invalidrequestwasreceived)
[unexpectedErrorProcessingRequest](https://www.apollographql.com/docs/apollo-server/integrations/plugins-event-reference/#unexpectederrorprocessingrequest)


![Screenshot_20230511_162124](https://github.com/hitz-group/apollo-prometheus-exporter/assets/89755878/0933a195-04e6-49f2-911c-e40c969d9253)

Reply:
Did we do a share screen session which really test e.g. a duplicate product validation - which shows up in the metrics like @quangthanh0909 mentioned ?

Do you guys know how to run the metrics server locally?

Reply:
I'm recoding to show what @quangthanh0909 mentioned

Reply:
a @hoang-tillpos https://www.loom.com/share/d108a49a18374f45a3429c44e493f1e3

Reply:
> 

Hi a @hoang-tillpos , we had a short time building this, and most of the code was reused from Van's works.
We didn't build this from the beginning.  We can have room for improvement to take advantage of the default plugin hook after the first release. 

What do you think ?

Reply:
I inherited from @vankongv